Core/Creatures: Drop difficulty entries

meji46 opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Dropped difficulty entries and moved all difficulty dependent fields to creature_template_scaling
   -  HealthScalingExpansion, HealthModifier, ManaModifier, ArmorModifier, DamageModifier, CreatureDifficultyID
-  Added difficulty field for table creature_questitem
-  Support for fallback difficulty in case of no data for current difficulty 

**Issues addressed:**
None


**Tests performed:**
Builds, tested in-game.


**Known issues and TODO list:** (add/remove lines as needed)

- In the future, loot should be linked to CreatureDifficultyID
- The query caching system requires a revamp in order to cache the difficulty dependent data. I have simplified it to only cache data for DIFFICULTY_NONE


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

meji46 committed:
Core/Creatures: Drop difficulty entries

meji46 committed:
Update

meji46 committed:
Fix nopch

Shauren commented on src/server/game/Entities/Creature/Creature.h:
```diff
@@ -469,7 +469,6 @@ class TC_GAME_API Creature : public Unit, public GridObject<Creature>, public Ma
 
         bool DisableReputationGain;
 
-        CreatureTemplate const* m_creatureInfo;                 // Can differ from sObjectMgr->GetCreatureTemplate(GetEntry()) in difficulty mode > 0
```
why remove the cached pointer?

meji46 replied:
I assumed that it was only used to identify difficulty entry https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Entities/Creature/Creature.cpp#L507

meji46 replied:
Fixed

Shauren commented:
Do these sql changes cover all creature templates that have difficulty_entry set?

meji46 commented:
> Do these sql changes cover all creature templates that have difficulty_entry set?

Yes, it moves the data from the dropped columns (base entry) to all scaling records. Then, the data from each difficulty is moved to the specific scaling record (or a default one is generated if missing)
![image](https://user-images.githubusercontent.com/2695278/234122116-1cc03892-5447-4a98-aed1-11658c8acbfa.png)
DifficultyId for normal mode is wrong (should be 3) but we preserve all modifiers

meji46 committed:
Restore cached pointer

meji46 committed:
Merge master

meji46 committed:
Update PR

Rename creature_template_scaling to creature_template_difficulty

Move Gold and TypeFlags fields to creature_template_difficulty

Update missing creature difficulties

meji46 committed:
Auras by difficulty

meji46 committed:
Missing override

meji46 committed:
Loot by difficulty

meji46 committed:
Merge master

Shauren committed:
Merge branch 'master' into meji46/creature-difficulty-entries

Shauren committed:
prepare merge

Shauren committed:
prepare merge 2
